### PR TITLE
Yotam playg 37 support insert

### PR DESCRIPTION
**Pull request by yotam-yakovw**
added 2 funcs for get and append requests to the spreadsheet api that are used by relevant methods for reading/writing data to the spreadsheet, also added func for getting the spreadsheet values resource to avoid repeating code.

### CODE REVIEW

File: data_loader/training_extractor.py [hunk 1]
```diff
@@ -12,20 +12,53 @@
 
 # The ID and range of a sample spreadsheet.
 SAMPLE_SPREADSHEET_ID = '1NGRUyzImlaUd-UrTkNXSd1JB7OOHnw6e1h4AohRXFK8'
+TEST_INSERT_SPREADSHEET_ID = '1KQ2nFO9oeDLhHmKh9UTvNnhepKhMuTOyKmh7Nnxbj80'
```
Comment (smartFreshBear) [inline data_loader/training_extractor.py:15]: That should be in a test file that tests your logic.
(we don't mix production code, and test code in the same file)

**Review by smartFreshBear — CHANGES_REQUESTED** [review]
good start, few points and questions:

1- How did you test that it is working?
2- The conventional name for it can be get_values, insert_values OR get_rows/insert_rows.
the name you gave is long and raises too many questions
2-We need clear speration between spreadsheet related logic AND business related logic.
Namely, we want one file-class(= let's say, TrainingStorer) that knows how to store training and it USES a-file class (dao = data access object, lets call it SpreadSheedDao)  that  would be a file that read and insert rows from spread sheet.

in the future, we might want to replace the SpreadSheedDao to use other technologcy like Postgress for exampl (or redis or mongo) , and it would be easy, because we dont need to touch TrainingStorer!

hope I clarify, PLEASE feel free to ask and get more hints

**Comment by yotam-yakovw** [comment]
Fixes:
- Training API calls and data handling separated to different files
- New files and functions got new clearer names (including one typo), changes needed to be updated in 'main_flow.py' and '__init__.py' files to use the correct function names
- Added a testing module for getting and inserting data to the spreadsheet, testing was based on testing files already in the project, testing variables moved to the new test file

Notes:
- Parameter handling for inserting data to spreadsheet was changed because it (seems) unnecessarily complicated and made a mess for the request. If the structure of the sent parameter should be changed I would like to get an explanation about how exactly it is sent
- About the testing module I'm not fully sure it is how it should be, it's made from what I managed to find online and existing test files I could find in the project, I'm sending it now as it is and expecting feedback
- Copied the token.pickle file from data_loader folder to the new test folder, just pointing that out to avoid mistakes

**Comment by smartFreshBear** [comment]
image
  1- let's talk a little about your commits - you almost doing it good.
 each commit suppose to pack an agenda of what you intended to do, which you did BUT
 take in account, that changes on changes you did, should be avoided as much as possible.
 for example:
 commit 1 = 'adding feature to made car faster'
 commit 2 = 'testing car is fast'
commit 3 = 'fixing some stuff I discovered later after testing car'

will just pollute the master branch.
to avoid it before updating PR we usually squash commit together (in my example 1 and 3)

 2- u have a conflict fix it when u have time :) 
 
 (other comments are on they way)

File: data_loader/spreadsheet_dao.py [hunk 2]
```diff
@@ -18,27 +17,7 @@
 """ TODO address + sheets from training set"""
```
Comment (smartFreshBear) [inline data_loader/spreadsheet_dao.py:17]: not yours, but delete

File: data_loader/test/test_training_interface_logic.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    def test_get_values_not_empty(self):
+        values = get_values()
+        assert len(values) != 0, 'Function returned an empty list'
+
+    def test_insert_values(self):
+        cells = [['a', 'b', 'c']]
+
+        response = insert_values(cells, TEST_INSERT_SPREADSHEET_ID)
+
+        assert response, 'Request returned bad response'
```
Comment (smartFreshBear) [inline data_loader/test/test_training_interface_logic.py:22]: good start here usually when you test dao you want to:
1- insert, verify what you inserted was...inserted.
2- delete what you inserted, verify it was deleted.

what I see is that you mostly check no exception was raised with some  very basic sanity testing.

File: data_loader/spreadsheet_dao.py [hunk 4]
```diff
@@ -55,3 +34,33 @@ def get_clinet_to_training_set():
...
+
+def get_values_resource():
+    service = get_client_to_training_set()
+    sheet = service.spreadsheets()
+    values = sheet.values()
+
+    return values
+
+
+def request_to_get_spreadsheet_values():
```
Comment (smartFreshBear) [inline data_loader/spreadsheet_dao.py:47]: get_spread_sheet_values would be more straightforward

File: data_loader/spreadsheet_dao.py [hunk 5]
```diff
@@ -55,3 +34,33 @@ def get_clinet_to_training_set():
...
+    resource = get_values_resource()
+    result = resource.get(spreadsheetId=SAMPLE_SPREADSHEET_ID,
+                          range=SAMPLE_RANGE_NAME).execute()
+    values = result.get('values', [])
+    return values
+
+
+def request_to_append_spreadsheet_values(values, target_spreadsheet):
```
Comment (smartFreshBear) [inline data_loader/spreadsheet_dao.py:55]: get_spread_sheet_values would be more straightforward

File: data_loader/training_storer.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import logging
+
+from data_loader.spreadsheet_dao import *
+
```
Comment (smartFreshBear) [inline data_loader/training_storer.py:4]: training_data_service, training_service would be the more conventional name (agreed in the industry)

so... can have training_service that uses spread_sheet_dao AND in 6 month we can make it uses some futuristic postgres_dao...

BUTTTT!
[this part is not a must, but bare with me if it trigger your curiosity!]
for this to happen we need to use interfaces, and classes

can you think of a way to design this?

**Review by smartFreshBear — CHANGES_REQUESTED** [review]
very good work on the fixes, please go over the comments!
good night!

File: data_loader/training_storer.py (hunk 6 above)
Comment (yotam-yakovw) [reply data_loader/training_storer.py:4]: well my curiosity is triggered but I would like some more elaborate explanation about this.
not sure about the interfaces part yet.

**Comment by yotam-yakovw** [comment]
fixed everything I could, would like to hear your feedback about the rewrite of the training functions and if there is a way to improve (probably...)
also waiting for answer on last conversation

**Review by smartFreshBear — DISMISSED** [review]
good work there please make sure u testing it also manually - means you delete the weights and train the model from the beginning (see nothing got broken with your new change)

File: stemming/stemming.py [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import copy
 import time
 import logging
-import pickle5 as pickle
+import pickle
 import requests
```
Comment (smartFreshBear) [inline stemming/stemming.py:5]: this suppose to be stashed (it served you local env right? we dont want it to go to prod server)

**Pull request closed without merging**